From 854499acfc4dc309d4ff80962eb7d7d909fac899 Mon Sep 17 00:00:00 2001
From: Wadamzmail <alksndrmetwakil@gmail.com>
Date: Sat, 25 Jul 2026 17:06:11 +0200
Subject: [PATCH 1/9] fix: use application class loader to find jrt:/ file
 system

The JRT file system is not available on Android. It needs to be part of the application.
To ensure Kotlin's CoreJrtFileSystem can find the jrt:/ file system, we need to make it
use the application class loader. This makes it possible for the consuming application
to provide its own implementation of the jrt:/ file system (as we do in Code On the Go).

---
 .../cli/jvm/modules/CoreJrtFileSystem.kt      | 31 ++++++++-----------
 1 file changed, 13 insertions(+), 18 deletions(-)

diff --git a/compiler/cli/cli-base/src/org/jetbrains/kotlin/cli/jvm/modules/CoreJrtFileSystem.kt b/compiler/cli/cli-base/src/org/jetbrains/kotlin/cli/jvm/modules/CoreJrtFileSystem.kt
index 335a4db..e075e42 100644
--- a/compiler/cli/cli-base/src/org/jetbrains/kotlin/cli/jvm/modules/CoreJrtFileSystem.kt
+++ b/compiler/cli/cli-base/src/org/jetbrains/kotlin/cli/jvm/modules/CoreJrtFileSystem.kt
@@ -23,7 +23,6 @@ import com.intellij.util.containers.ConcurrentFactoryMap
 import com.intellij.util.io.URLUtil
 import java.io.File
 import java.net.URI
-import java.net.URLClassLoader
 import java.nio.file.FileSystem
 import java.nio.file.FileSystems
 
@@ -73,24 +72,20 @@ class CoreJrtFileSystem : DeprecatedVirtualFileSystem() {
 
         private val globalJrtFsCache = ConcurrentFactoryMap.createMap<String, FileSystem?> { jdkHomePath ->
             val jdkHome = File(jdkHomePath)
-            val jrtFsJar = loadJrtFsJar(jdkHome) ?: return@createMap null
+            loadJrtFsJar(jdkHome) ?: return@createMap null
             val rootUri = URI.create(StandardFileSystems.JRT_PROTOCOL + ":/")
+
             /*
-              The ClassLoader, that was used to load JRT FS Provider actually lives as long as current thread due to ThreadLocal leak in jrt-fs,
-              See https://bugs.openjdk.java.net/browse/JDK-8260621
-              So that cache allows us to avoid creating too many classloaders for same JDK and reduce severity of that leak
-            */
-            if (isAtLeastJava9()) {
-                // If the runtime JDK is set to 9+ it has JrtFileSystemProvider,
-                // but to load proper jrt-fs (one that is pointed by jdkHome) we should provide "java.home" path
-                FileSystems.newFileSystem(rootUri, mapOf("java.home" to jdkHome.absolutePath))
-            } else {
-                val classLoader = URLClassLoader(arrayOf(jrtFsJar.toURI().toURL()), null)
-                // If the runtime JDK is set to <9, there are no JrtFileSystemProvider,
-                // we should create classloader with jrt-fs.jar, and DO NOT NEED to pass "java.home" path,
-                // as otherwise it will incur additional classloader creation
-                FileSystems.newFileSystem(rootUri, emptyMap<String, Nothing>(), classLoader)
-            }
+             * CodeOnTheGo bundles its own jrt-fs implementation inside the APK.
+             * Therefore, we do not need to check the runtime Java version here.
+             * Always resolve the JRT FileSystemProvider from the application's ClassLoader,
+             * using the bundled implementation regardless of the Android or Java runtime.
+             */
+            FileSystems.newFileSystem(
+                rootUri,
+                mapOf("java.home" to jdkHome.absolutePath),
+                CoreJrtFileSystem::class.java.classLoader
+            )
         }
     }
-}
+}
\ No newline at end of file
-- 
2.52.0

